### DISCUSSION TITLE
How to access .env variables outside of <script setup> or lifecycle hooks?

### DISCUSSION THREAD

**Discussion by pschnyder** [nuxt/nuxt#16098] [Questions] [answered]
Hello,

Is there a way to access .env variables outside of `<script setup>`  or Lifecycle Hooks?

In my Nuxt 3 application, I have a few composables that make API calls to the a server that has a different URL than the frontend (i.e. the Nuxt app and server are fully decoupled). As I'm often using these composables outside of <script setup> or lifecycle hooks, I can't use `useRuntimeConfig()` and `process.env.BACKEND_URL_BASE` doesn't work. Is there a solution to this? Right now I'm just passing the baseURL as a parameter to the composable but this doesn't feel right...

**.env file**
```
# URL
BACKEND_URL_BASE="http://localhost:8000"
```

**nuxt.config.ts**

```
import { defineNuxtConfig } from 'nuxt'`
export default defineNuxtConfig({
    runtimeConfig: {
        public: {
            backendUrlBase: process.env.BACKEND_URL_BASE,
          }
    }
})`
```
That's how I'm trying to access the .env file in the composable:
```
import { FetchOptions } from "ohmyfetch";

export const useMyComposable = async () => {

  const myApiCall = await $fetch(`${process.env.BACKEND_URL_BASE}/something`, {
    method: 'GET'
  }).then((data) => {
    //do something
  }).catch((error) => {
    console.log(error);
   }
  )

[...]
};
```

Thanks so much for your help!

**Comment by leonardssh** [comment, marked as answer]
https://github.com/nuxt/framework/issues/5254#issuecomment-1145922571

**Reply by pschnyder** [reply]
Thanks! So the `process.env.BACKEND_URL_BASE` does not work because it is in a composable?

**Reply by leonardssh** [reply]
Not because it's composable, you can't use `process.env` in the client, if you try to access it in the server `server/api/hello.ts` you'll see it works.

```ts
// server/api/hello.ts

const config = useRuntimeConfig();

export default (req, res) => {
  console.log(config.public.apiBase);
  // or
  console.log(process.env.NUXT_PUBLIC_API_BASE);
};
```

**Reply by leonardssh** [reply]
For now, you can use shtse8's approach: https://github.com/nuxt/framework/discussions/5886#discussioncomment-3140447

**Reply by pschnyder** [reply]
Ok! thanks so much for the help!

**Comment by shtse8** [comment]
I need this feature too.
I have some scripts running as services outside Nuxt.
I want to get the runtime config of Nuxt. Any helper method for us to get the runtime config?

Below is my polyfill to get the nuxt runtime config. it is better if nuxt can provide us.
```ts
import fs from 'fs'
import type { RuntimeConfig } from '@nuxt/schema'
import { snakeCase } from 'lodash-es'
import * as dotenv from 'dotenv'
import config from '~/nuxt.config'
import { Lazy } from '~/utils'

class ConfigReader {
  private _config: RuntimeConfig | undefined
  constructor() {
  }

  get config(): RuntimeConfig {
    if (!this._config)
      this._config = this.get()

    return this._config
  }

  private get(): RuntimeConfig {
    const envFileStr = fs.readFileSync('.env', { encoding: 'utf8' })
    const envFile = dotenv.parse(envFileStr)
    const env = Object.assign({}, process.env, envFile)
    if (config.runtimeConfig)
      this.fill(env, config.runtimeConfig, 'NUXT_')
    return config.runtimeConfig as RuntimeConfig
  }

  private fill(env: Record<string, any>, obj: Record<string, any>, prefix = '') {
    for (const key in obj) {
      const name = prefix + snakeCase(key).toUpperCase()
      if (typeof obj[key] === 'object')
        this.fill(env, obj[key], `${name}_`)
      else if (name in env)
        obj[key] = env[name]
    }
  }
}

const configMock = new Lazy(() => new ConfigReader())
export function useConfig() {
  try {
    return useRuntimeConfig()
  }
  catch (e) {
    return configMock.value.config
  }
}
```

**Reply by leonardssh** [reply]
Interesting approach, good job 👍. I'd really like to see it provided by Nuxt too.

**Comment by chairuosen** [comment]
same here.
`useXXX` api is more like a view-level api just for rendering.
env config api should be more low level and globally.

`useRuntimeConfig` is not a good api design
